src/main/java/com/troblecodings/launcher/util/ConnectionUtil.java

```diff
@@ -1,121 +1,172 @@
 
 	// Checks if the file exist and that its sha1 hash equals the given
 	// returns true if all the checks pass
-	public static boolean validate(String name, String sha1) throws Throwable {
-		MessageDigest digest = MessageDigest.getInstance("SHA-1");
-		Path pathtofile = Paths.get(name);
-		if (!Files.exists(pathtofile))
-			return false;
-		try (DigestInputStream stream = new DigestInputStream(Files.newInputStream(pathtofile), digest)) {
-			while (true) {
-				byte[] buffer = new byte[8192];
-				if (stream.read(buffer, 0, 8192) <= 0)
-					break;
+	public static boolean validate(String name, String sha1) {
+		try {
+			MessageDigest digest = MessageDigest.getInstance("SHA-1");
+			Path pathtofile = Paths.get(name);
+			if (!Files.exists(pathtofile))
+				return false;
+			try (DigestInputStream stream = new DigestInputStream(Files.newInputStream(pathtofile), digest)) {
+				while (true) {
+					byte[] buffer = new byte[8192];
+					if (stream.read(buffer, 0, 8192) <= 0)
+						break;
+				}
+				byte[] digestreturn = digest.digest();
+				BigInteger sha1bigintegers = new BigInteger(1, digestreturn);
+				String sha1result = sha1bigintegers.toString(16);
+				while (sha1result.length() < 40) {
+					sha1result = "0" + sha1result;
+				}
+				return sha1result.equalsIgnoreCase(sha1);
+			} catch (IOException e) {
+				e.printStackTrace();
 			}
-			byte[] digestreturn = digest.digest();
-			BigInteger sha1bigintegers = new BigInteger(1, digestreturn);
-			String sha1result = sha1bigintegers.toString(16);
-			while (sha1result.length() < 40) {
-				sha1result = "0" + sha1result;
-			}
-			return sha1result.equalsIgnoreCase(sha1);
+		} catch (NoSuchAlgorithmException e) {
+			e.printStackTrace();
 		}
+		return false;
 	}
 
 	// This attempts to download a file if it isn't valid
-	public static void validateDownloadRetry(String url, String name, String sha1) throws Throwable {
-		byte times = 0;
-		if (url.isEmpty()) {
-			if (!ConnectionUtil.validate(name, sha1))
-				throw new VerifyError("Couldn't verify " + name);
-			return;
-		}
-		while (!ConnectionUtil.validate(name, sha1)) {
-			ConnectionUtil.download(url, name);
-			if (times == 5) {
-				ErrorDialog.createDialog(new VerifyError("Couldn't verify " + name));
-				break;
+	public static void validateDownloadRetry(String url, String name, String sha1) {
+		executors.submit(() -> {
+			byte times = 0;
+			if (url.isEmpty()) {
+				if (!ConnectionUtil.validate(name, sha1))
+					throw new VerifyError("Couldn't verify " + name);
+				return;
 			}
-			times++;
-		}
+			while (!ConnectionUtil.validate(name, sha1)) {
+				if (times == 3) {
+					Launcher.INSTANCEL.setPart(new ErrorPart(Launcher.INSTANCEL.getPart(), "Error verifying " + Paths.get(name).getFileName().toString(), "The file failed to download correctly after 3 tries!"));
+					break;
+				}
+				ConnectionUtil.download(url, name);
+				times++;
+			}
+		});
 	}
 
 }
```

Comment:
I can't find out why, but the code throws exceptions, if the modpack is started after using the reset feature. This happend with different files, but all with the same exception.
```
java.nio.file.NoSuchFileException: C:\Users\andso\AppData\Roaming\gir\libraries\net\sf\jopt-simple\jopt-simple\5.0.3\jopt-simple-5.0.3.jar.tmp
	at sun.nio.fs.WindowsException.translateToIOException(WindowsException.java:79)
	at sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:97)
	at sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:102)
	at sun.nio.fs.WindowsFileCopy.move(WindowsFileCopy.java:318)
	at sun.nio.fs.WindowsFileSystemProvider.move(WindowsFileSystemProvider.java:287)
	at java.nio.file.Files.move(Files.java:1395)
	at com.troblecodings.launcher.util.ConnectionUtil.download(ConnectionUtil.java:113)
	at com.troblecodings.launcher.util.ConnectionUtil.download(ConnectionUtil.java:90)
	at com.troblecodings.launcher.util.ConnectionUtil.lambda$0(ConnectionUtil.java:166)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

Reply:
Very good testing 👍 I will take a look into it

Reply:
This is an issue with the executor pool ... turns out parallel downloading isn't the greatest of all ideas. I have removed the feature. Thanks for checking.